### ISSUE TITLE
[New Sensor]:  Kpro Beacon

### ISSUE THREAD

**Issue by barkdadof4** [custom-components/ble_monitor#835]
### Sensor Description

KKM K6

### Additional information

Mac address dd340206d06c kpro

Here's their iOS SDK: https://github.com/kkmhogen/kbeaconlib2
Specifically, check out this file for parsing the advertisement packet w/ temp and humidity: https://github.com/kkmhogen/kbeaconlib2/blob/df34f4bfeb80d9793e098f3e5b57382bf6a38ea4/kbeaconlib2/Classes/KBAdvPacketSensor.swift

I am not smart enough to write the python code from the above information
[dump.txt](https://github.com/custom-components/ble_monitor/files/8514482/dump.txt)
.

### BLE advertisements

```shell
04 3E 24 02 01 04 01 6C D0 06 02 34 DD 18 09 16 80 20 6D 01 
  00 00 00 00 0D 09 4B 42 50 72 6F 5F 32 31 34 38 31 30 D3
```

**Comment by barkdadof4** [comment]
Sorry pushed the wrong button

**Comment by Ernst79** [comment]
hmmm, I also found a google UUID (and length). Normally, UUID is a fixed value according to the Bluetooth specs. But this sensor seems to be sending two UUID's

**Google UUID**
`1216aafe21010f0e07192a224ffffcffec03eb`

12 = length
16 = manufacturer specific data
aafe = UUID (0xFEAA = Google LLC)
21010f0e07192a224ffffcffec03eb = data

**Unknown UUID**
`091680206d0100000000` 

`09` = length
`16` = manufacturer specific data
`8020` = UUID (not officially registered at the Bluetooth organization)
`6d0100000000` = data

**Comment by Ernst79** [comment]
I'm trying to make some sense out of it, but still failing. 

If I understand the link you gave correctly, we have

version 1 byte? 
battery level/voltage 2 bytes
temperature 2 bytes
humidity 2 bytes
acceleration x-axis 2 bytes
acceleration y-axis 2 bytes
acceleration z-axis 2 bytes

`1216AAFE 2101 0F0E 0719 3022 CBFF F5FF EF03 FB C6`
`1216AAFE 2101 0F0E 0719 2A22 4FFF FCFF EC03 EB D3`
`1216AAFE 2101 0F0E 0719 3022 CBFF F9FF E803 E9 D1`

Bytes normally have to be read backwards, so `0F0E` has to be read as `0x0E0F`, after conversion from hex to dec, this is equal to 3599. But I can't make sense of the numbers, they don't seem to match with 2125 (`0x084D`) or 212 (`0x00D4)` for example. 

In the link you gave, I don't see any conversion (integers are always numbers without digits, so normally there is some formula somewhere, e.g. temperature / 100, to get the number of digits.

**Comment by barkdadof4** [comment]
I have some other docs to share
![KproMap](https://user-images.githubusercontent.com/13120563/164333625-4274053a-8989-4738-a327-25eb32aaf826.jpg)
[KBeaconPro download resource link (2).pdf](https://github.com/custom-components/ble_monitor/files/8525781/KBeaconPro.download.resource.link.2.pdf)
[K6 (1).pdf](https://github.com/custom-components/ble_monitor/files/8525788/K6.1.pdf)
.

This device also has the ability to advertise with different protocols. 

Thanks

**Comment by Ernst79** [comment]
Thanks, following your link, I get

`1216AAFE 21 01 0F 0E071930 22CB FFF9 FFE8 03E9 D1`

`12` = length
`16`= adv type manufacture specific data
`AAFE` = UUID (google llc)
`21` = version
`01` = version tag (always 01)
`0F` = Sensor Mask
`0E07` = voltage = 3591 mV
`1930` = temperature = 25.48
`22CB` = humidity = 34.203
`FFF9` = acceleration x = -7 mG
`FFE8` = acceleration x = -24 mG
`03E9` = acceleration z = 1001 mG
`D1` = ???

Temperature is a little higher than you said, but I guess this is ok.

**Comment by barkdadof4** [comment]
Great thanks can you share the math used to get Temp and humidity.  I seem to be missing something.

**Comment by Ernst79** [comment]
first byte is the number before the point, second byte is number after the point. 

So, `1930` 
--> `19` in hex is 25 in dec
--> `30` in hex = 48 in dec
--> 25.48

**Comment by barkdadof4** [comment]
My bad I was not doing them individually. I was trying to convert 1930.

Again thanks for your help.

**Comment by Ernst79** [comment]
I was too in my fist attempt, that is how it normally is done, but not for this sensor (actually, only temperature and humidity are using this strange format).

**Comment by Ernst79** [comment]
The PR with the parser code is in #840

I'll add it to ble_parser / BLE monitor next week, I first have to move to my new house.

**Comment by barkdadof4** [comment]
Ok enjoy your new house.  I am also learning that all advertisement packets are not created equal.  

Trying to understand the code better in the ThermoBeacon.py and ThermoBeacon2.py code,  The author does a lot of manipulation to get to the number.  So I tried to make it follow what you have showed me and of course it does not work.  Just ways to confuse the amateurs. This is also why you have such a hobby and need to create and adjust libraries all the time.

Again thanks for the knowledge transfer.

**Comment by Ernst79** [comment]
I have added this sensor to both BLE monitor (8.6.2) and bleparser (3.1.0). 

If you have questions how it works, let me know.  Most of the "logic" is in the new kkm.py file.

**Comment by DigiH** [comment]
@Ernst79 as we also added some KKM sensors I had a look at your implementation, and found the following.

With the fixed point 8.8 format for temperature and humidity, the decimal part looks to me like binary fractions, not just hex to decimal conversion. Otherwise any hex value equal or higher than 64 would create a new integer part.

Also for the negative temperatures you having

temperature = -(temp + 128 + temp_frac / 100)

makes things very chilly, e.g. for

**ff8e** = -(-1 + 128 + 142/100) = **-128.42°C** with your calculation

whereas I'm pretty certain it should be **-0.4453125°C**

Also for the positive temperatures I added your test case as well, getting different result for the decimal parts.
